### PR TITLE
Alexmsettle parse callid

### PR DESCRIPTION
**Pull request by alexmsettle**
I updated the parse and prof scripts to include the callid and input_callid fields in the final report.  The callid is the unique id of the op, and the input_callids are the inputs of the predecessor ops.  This is sufficient information to build the network graph.  Software can postprocess the pyprof report and produce a graph visualization.

`
Idx | SeqId | AltSeqId | callid | input_callids | TId
-- | -- | -- | -- | -- | --
1 | 4848 | - | [2] | [-1] | 424
2 | 4849 | - | [6] | [-1] | 424
3 | 4849 | - | [15] | ['6', '2'] | 424
4 | 4849 | - | [15] | ['6', '2'] | 424
5 | 4850 | - | [19] | [-1] | 424
6 | 4850 | - | [20] | ['15', -1] | 424
7 | 4851 | - | [22] | ['20'] | 424
8 | 4852 | - | [24] | ['22'] | 424
9 | 4853 | - | [28] | [-1] | 424
10 | 4854 | - | [40] | ['24', '28'] | 424
11 | 4855 | - | [44] | [-1] | 424
12 | 4855 | - | [45] | ['40', -1] | 424

`

### CODE REVIEW

File: pyprof/parse/kernel.py [hunk 1]
```diff
@@ -121,7 +124,8 @@ def setRunTimeInfo(self, info):
         self.tid = info['tid']
         self.objId = info['objId']
         assert (self.rStartTime < self.rEndTime)
-        assert (self.rStartTime < self.kStartTime)
+        if self.kStartTime:
+            assert (self.rStartTime < self.kStartTime)
```
Comment (adityaiitb) [inline pyprof/parse/kernel.py:128]: While this is practically true, I vaguely remember having this assertion and it getting triggered. I did not investigate it but it could be due to very slight differences in the CPU and GPU timestamps (unix timestamps of whatever they use).

File: pyprof/parse/nsight.py [hunk 2]
```diff
@@ -224,7 +248,7 @@ def filterTrace(mlist):
             if m.find("CheckpointFunctionBackward") >= 0:
                 continue
 
-            if ("_backward, seq =" in m) or ("Backward, seq =" in m) or ("Backward0, seq =" in m):
+            if re.search(r"_backward, seq =|Backward[0-9]*, seq=", m):
```
Comment (adityaiitb) [inline pyprof/parse/nsight.py:251]: This is very delicate code. A space is missing between `seq=` which will break a lot of code (probably not because the code is resilient) but will result in incorrect correlation. The current QA of lenet is probably not effective. Can we move any refactoring to a different PR, so that its easy to roll back if errors arise.

File: pyprof/prof/convert.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ def __init__(self, d):
 
         assert (mod == "Tensor")
         assert (op in Convert.ops)
-        assert (len(args) == 1)
+        assert (len(args) >= 1), "Args {} Len {}".format(args, len(args))
```
Comment (adityaiitb) [inline pyprof/prof/convert.py:39]: Can we move this to a separate PR. Was there a reason to weaken the assertion? If the assertion indeed got triggered, we should fix it as a separate PR.

File: pyprof/prof/misc.py [hunk 4]
```diff
@@ -129,7 +129,7 @@ def __init__(self, d):
         self.op_ = op
         self.args = args
 
-        assert (mod == "Tensor")
+        assert (mod == "Tensor" or mod == 'torch'), "Mod {} marker {}".format(mod, marker)
```
Comment (adityaiitb) [inline pyprof/prof/misc.py:132]: Separate PR.

File: pyprof/prof/index_slice_join_mutate.py [hunk 5]
```diff
@@ -139,11 +139,7 @@ def __init__(self, d):
         args = list(filter(lambda x: x['name'] != 'out', args))
         assert (len(args) == 3)
 
-        #Get input
-        if (args[0]['name'] == ""):
-            arg = args[0]
-        else:
-            arg = list(filter(lambda x: x['name'] == "input", args))[0]
+        arg = args[0]
```
Comment (adityaiitb) [inline pyprof/prof/index_slice_join_mutate.py:142]: For a function `foo(a, b, c)`, a user can call it as `foo(c=20, b=10, a=5)`. If we want `a` we cannot assume the arguments will appear in the order of the function signature and always pick the first one.

File: pyprof/prof/pointwise.py [hunk 6]
```diff
@@ -145,13 +149,13 @@ def bytes_flops(self):
 
         # Unary
         if self.op() in Pointwise.unary + Pointwise.representation:
-            assert (len(self.input) == 1)
+            assert (len(self.input) >= 1)
```
Comment (adityaiitb) [inline pyprof/prof/pointwise.py:152]: Reason for weakening the assertion?

File: pyprof/prof/pointwise.py [hunk 7]
```diff
@@ -145,13 +149,13 @@ def bytes_flops(self):
...
             b = 2 * self.input[0].bytes
             f = self.input[0].size
 
         elif self.op() in Pointwise.exp_log + Pointwise.trig_trans + \
                 Pointwise.sqrt + Pointwise.error:
-            assert (len(self.input) == 1)
+            assert (len(self.input) >= 1)
```
Comment (adityaiitb) [inline pyprof/prof/pointwise.py:158]: Reason for weakening the assertion?

File: pyprof/prof/pointwise.py [hunk 8]
```diff
@@ -182,9 +187,9 @@ def bytes_flops(self):
                     assert False, e
 
         elif self.op() in Pointwise.power:
-            assert (len(self.input) == 2)
-            out = Tensor.broadcast(self.input)
-            b = reduce(operator.add, [t.bytes for t in self.input])
+            assert (len(self.input) >= 2)
+            out = Tensor.broadcast(self.input[0:2])
+            b = reduce(operator.add, [t.bytes for t in self.input[0:2]])
```
Comment (adityaiitb) [inline pyprof/prof/pointwise.py:192]: Same as above.

File: pyprof/prof/pointwise.py [hunk 9]
```diff
@@ -160,11 +164,12 @@ def bytes_flops(self):
                 Pointwise.binary + Pointwise.ibinary + \
                 Pointwise.logical + Pointwise.ilogical:
 
-            assert (len(self.input) == 2)
-            out = Tensor.broadcast(self.input)
+            assert (len(self.input) >= 2), "Op {} inputs {} unfiltered args {}"\
+                    .format(self.op(), self.input, self.unfiltered_args)
+            out = Tensor.broadcast(self.input[0:2])
```
Comment (adityaiitb) [inline pyprof/prof/pointwise.py:169]: Same as above.

File: pyprof/prof/reduction.py [hunk 10]
```diff
@@ -35,9 +35,6 @@ def __init__(self, d):
         assert (mod in ["torch", "Tensor"])
         assert (op == "mean")
 
-        #Filter out named parameters
-        args = list(filter(lambda x: x['name'] == '', args))
-
```
Comment (adityaiitb) [inline pyprof/prof/reduction.py:40]: https://pytorch.org/docs/stable/_modules/torch/nn/functional.html#softmax

I have tested the code on a lot of esoteric code which use named arguments. What is the reason for removing this?

File: pyprof/prof/reduction.py [hunk 11]
```diff
@@ -93,11 +90,7 @@ def __init__(self, d):
         assert (op == "sum")
         assert (len(args) >= 1)
 
-        #Get input
-        if (args[0]['name'] == ""):
-            i = args[0]
-        else:
-            i = list(filter(lambda x: x['name'] == "input", args))[0]
+        i = args[0]
```
Comment (adityaiitb) [inline pyprof/prof/reduction.py:93]: Same as above. We cannot assume the arguments appear in the order specified by the API documentation.

File: pyprof/prof/softmax.py [hunk 12]
```diff
@@ -86,16 +86,11 @@ def __init__(self, d):
         assert (mod in ["torch", "torch.nn.functional"])
         assert (op == "log_softmax")
 
-        #Filter out named parameters
-        args = list(filter(lambda x: x['name'] == '', args))
```
Comment (adityaiitb) [inline pyprof/prof/softmax.py:90]: See above.

File: pyprof/prof/softmax.py [hunk 13]
```diff
@@ -86,16 +86,11 @@ def __init__(self, d):
...
 
         assert (len(args) <= 2)
 
         #Get input
-        if (args[0]['name'] == ""):
-            i = args[0]
-        else:
-            i = list(filter(lambda x: x['name'] == "input", args))[0]
```
Comment (adityaiitb) [inline pyprof/prof/softmax.py:98]: Do not remove. See above.

**Comment by adityaiitb** [comment]
- Can we convert the values inside `input_callids` from string to integer.
- Can we restrict the PR to just the changes required for call id support.
- A lot of strict assertions and code resiliency was developed after testing on numerous models / libraries / PyTorch versions. Some code can appear unnecessary but it was added for a reason to handle all the weird things Python allows a user to do. Can we not remove/simplify code.

**Comment by adityaiitb** [comment]
I thought the graph call id was a dlprof only feature? If that is not the case, we should refactor the previous PR so that it is a stand alone stub/feature (as I had proposed earlier) and can be enabled by both pyprof and dlprof. Enabling graph information should not require enabling dlprof.

File: pyprof/parse/nsight.py (hunk 2 above)
Comment (adityaiitb) [reply pyprof/parse/nsight.py:251]: We should use a regex which matches the code i.e. `Backward` and `Backward0` as opposed to generalizing it to match 0 through 9. Reversing from a relaxed pattern to a stricter pattern will become impossible in the future for fear of breaking code.

**Comment by adityaiitb** [comment]
@debermudez @alexmsettle 
I just saw this comment `The 'name' field is being used to specify the unique name of the input node`. In the previous PR did we reuse the `name` field for something else? **That should not have been done.** It is typically empty, but it captures the name of the argument when the user chooses to do so e.g. `foo(input=20)`.

File: pyprof/parse/kernel.py (hunk 1 above)
Comment (alexmsettle) [reply pyprof/parse/kernel.py:128]: I have not seen this be an issue - I can change it to self.rStartTime <= self.kStartTime.  

Or I can remove it - but it looks to me like these scripts are biased to using assertions as much as possible

File: pyprof/parse/nsight.py (hunk 2 above)
Comment (alexmsettle) [reply pyprof/parse/nsight.py:251]: ok - so it sounds like there is not sufficient testing in place to support refactoring code?

I usually refactor as a I go.  I can put this back, but the chances of me later coming back to refactor it are pretty low.

File: pyprof/prof/convert.py (hunk 3 above)
Comment (alexmsettle) [reply pyprof/prof/convert.py:39]: This assertion was changed because the filtering of args is broken.

It appears that the software is trying to somehow keep track of keyword args and args, but it is not at all clear from any of the code that this is the case.

I think the code is assuming that args with the name key == "" means that it was not a keyword argument.  And args that have the name field set are keyword arguments.  For some reason that is also not clear from any of the code, the assumption is that unnamed args refer to the tensors and named args refer to hyper parameters.

I want the unique name of the ops to be passed down from nvtx markers so we can use them to identify the ops in the post processing stage.

File: pyprof/prof/misc.py (hunk 4 above)
Comment (alexmsettle) [reply pyprof/prof/misc.py:132]: Why?

File: pyprof/prof/index_slice_join_mutate.py (hunk 5 above)
Comment (alexmsettle) [reply pyprof/prof/index_slice_join_mutate.py:142]: same as above the arg filtering is broken - what are you trying to capture in the filter command?

File: pyprof/prof/pointwise.py (hunk 6 above)
Comment (alexmsettle) [reply pyprof/prof/pointwise.py:152]: Also - related to arg filtering problem.  As soon as args have a meaningful name the filtering breaks down.

File: pyprof/prof/pointwise.py (hunk 7 above)
Comment (alexmsettle) [reply pyprof/prof/pointwise.py:158]: same as above.  I only make changes to assertions when I hit errors in the code.  This is a problem with the undocumented assumption that inputs are partitioned into named and unnamed arguments and only the unnamed ones are useful to these scripts

File: pyprof/prof/pointwise.py (hunk 8 above)
Comment (alexmsettle) [reply pyprof/prof/pointwise.py:192]: same response as above

File: pyprof/prof/pointwise.py (hunk 9 above)
Comment (alexmsettle) [reply pyprof/prof/pointwise.py:169]: ditto

File: pyprof/prof/reduction.py (hunk 10 above)
Comment (alexmsettle) [reply pyprof/prof/reduction.py:40]: same as above

File: pyprof/prof/reduction.py (hunk 11 above)
Comment (alexmsettle) [reply pyprof/prof/reduction.py:93]: ditto

There needs to be better testing in this repo else it will be impossible to submit changes

File: pyprof/prof/softmax.py (hunk 12 above)
Comment (alexmsettle) [reply pyprof/prof/softmax.py:90]: ditto - bad assumption about keyword and non keyword args

File: pyprof/prof/softmax.py (hunk 13 above)
Comment (alexmsettle) [reply pyprof/prof/softmax.py:98]: same response as above

**Comment by alexmsettle** [comment]
Ok I need to know how to proceed with this.

It sounds like there is not sufficient testing for users to make changes to this code base.  If someone makes a change and the review states something like this code is fragile and it might break a user's test case, then that means the testing is not sufficient.

**Comment by adityaiitb** [comment]
As an overarching comment, which @dzier has suggested in the past, and I agree, is that a single PR should handle just one issue / bug / feature. Refactoring unrelated code, while highly important, should be part of a separate PR.

File: pyprof/parse/kernel.py (hunk 1 above)
Comment (adityaiitb) [reply pyprof/parse/kernel.py:128]: Yes, I have used as many assertions as possible. `self.rStartTime <= self.kStartTime` will have the same issue as I mentioned. We can move it to a separate PR and try it out on a variety of networks.

File: pyprof/parse/nsight.py (hunk 2 above)
Comment (adityaiitb) [reply pyprof/parse/nsight.py:251]: Yes, please put it back or move it to a separate PR. As you can imagine a lot of code (especially during backprop) is based on heuristics and it tries to decipher undocumented names/patterns used by the PyTorch backend. That's the part which is delicate. The purpose of hundreds of assertions is to catch errors when any of those assumptions no longer hold true. Adding a test for every heuristic can be done but requires a lot of effort.

File: pyprof/prof/convert.py (hunk 3 above)
Comment (adityaiitb) [reply pyprof/prof/convert.py:39]: That is the Python definition of `args` and `kwargs`. `args` are passed as a list (without names) and `kwargs` are passed as a dictionary (key-value pair). This fact was used when the NVTX marker was created. `name == ""` when its args and `name == foo` when its kwargs.

`args` and `kwargs` has no relationship to Pytorch / tensors / hyperparameters. It just depends on how the user decided to pass an argument to a function.

If you want unique name for ops, we can create an extra field in the NVTX marker. It should be really simple to do and does not require any hacking / overloading of existing keys in the marker.

File: pyprof/prof/misc.py (hunk 4 above)
Comment (adityaiitb) [reply pyprof/prof/misc.py:132]: @dzier suggestion of using 1 PR for one isolated issue/feature.

Comment (dzier) [reply pyprof/prof/misc.py:132]: Commit really, you have multiple commits per PR, just need to make sure we don't squash the commits.

**Comment by dzier** [comment]
The general comment I made before is that separate issues / refactoring should be in separate commits. But a PR can contain multiple commits. You can review this code commit by commit if you would like. Separate PR's just make it easier, but not necessary.

Alex, are you aware of how to add CI tests for our internal CI/CD? If there is a lack of tests, lets add them and make sure these changes are passing.

I would also like to point out that most of the nvmarker.py code will be cleaned up and most of the DLPROF specific code will be moved to a separate dlprof_nvmarker.py, or into the DLProf code base specifically. It may be better to apply these changes to the code base after it is cleaned up in the next couple of weeks.

File: pyprof/prof/index_slice_join_mutate.py (hunk 5 above)
Comment (adityaiitb) [reply pyprof/prof/index_slice_join_mutate.py:142]: Let me explain with an example. Consider the pytorch API for `relu`

```
torch.nn.functional.relu(input, inplace=False)
```

It can be called in a variety of ways e.g.
```
a = torch.ones(5,5)
relu(a)
relu(a, inplace=True)
relu(input=a, inplace=True)
relu(inplace=True, input=a)
```

The purpose of filtering is to determine `a` in all cases. Depending on how the function was called it can be part of `args` or `kwargs`. That's the intent, and its possible in a few cases the code is off. There is an assertion after the argument capture to verify if we got it right. In the past 2 years, they have never fired (or barely).

File: pyprof/prof/pointwise.py (hunk 6 above)
Comment (adityaiitb) [reply pyprof/prof/pointwise.py:152]: The problem is hacking / overloading an existing keyword (which identifies args and kwargs) for something else.

File: pyprof/prof/pointwise.py (hunk 7 above)
Comment (adityaiitb) [reply pyprof/prof/pointwise.py:158]: Same as above. By definition arguments in python are partitioned into a list and a dictionary and we handle both cases. Both are extremely useful. See [conv.py](https://github.com/NVIDIA/PyProf/blob/master/pyprof/prof/conv.py#L82-L108) for a really complicated example of how `name` is used. Just because its mostly empty does not mean its open to abuse.

File: pyprof/prof/pointwise.py (hunk 8 above)
Comment (adityaiitb) [reply pyprof/prof/pointwise.py:192]: Same as above. Please don't arbitrarily overload existing key-value pairs.

File: pyprof/prof/pointwise.py (hunk 9 above)
Comment (adityaiitb) [reply pyprof/prof/pointwise.py:169]: ditto

File: pyprof/prof/reduction.py (hunk 10 above)
Comment (adityaiitb) [reply pyprof/prof/reduction.py:40]: Same as above.

File: pyprof/prof/reduction.py (hunk 11 above)
Comment (adityaiitb) [reply pyprof/prof/reduction.py:93]: The source of the problem is not testing (we can of course improve on that) but the fact that the `name` keyword which has a  well thought out intent and purpose was overloaded and abused without consulting.

File: pyprof/prof/softmax.py (hunk 12 above)
Comment (adityaiitb) [reply pyprof/prof/softmax.py:90]: I can add a comment in the `nvmarker.py` file to more clearly specify what do the individual fields capture.

**Comment by adityaiitb** [comment]
New features (such as the current one) in a tool are extremely useful. I don't mean to preach but, in my experience it typically requires discussion / some sort of refactoring of the existing code so that the new feature is not a hack but looks like an integral part of the tool. Only then adding new features is sustainable.

**Comment by alexmsettle** [comment]
Well updating the 'name' field of the args is not something I did.  The changes I made related to kwargs and args were made because of this existing code.  Also worth noting is that there is no documentation that indicates the name field is special.  There are also no tests in place that check for the underlying assumption that name shall be empty for some case and not others.

git diff 36da7d4 0b54686 nvmarker.py

`-        cid = op_to_out_tensor_map.get(arg.data_ptr(), -1)
-        name = call_id_to_op_map.get(int(cid), "")
+        if config.capture_input_ops:
+            cid = dlprof.op_to_out_tensor_map.get(arg.data_ptr(), -1)
+            name = dlprof.call_id_to_op_map.get(int(cid), "")
`

`git  blame nvmarker.py | grep call_id_to_op_map
217:6b01ae09 (Elias Bermudez 2020-09-14 15:57:23 -0700 217)                 dlprof.call_id_to_op_map[dlprof.call_id] = traceMarker_dict['funcStack']
232:0b546866 (Elias Bermudez 2020-09-09 11:58:53 -0700 232)             name = dlprof.call_id_to_op_map.get(int(cid), "")`

**Comment by adityaiitb** [comment]
I see. @alexmsettle Thank you and I sincerely apologize for incorrectly assuming that the above change was introduced in the previous PR (as I mentioned here https://github.com/NVIDIA/PyProf/issues/132) and was the reason for the current assertion failures. I clearly remember having the same discussion with @debermudez about `args` and `kwargs` and why the `name` field is almost always empty but serves an important purpose. I am surprised the code still got checked in. The reason I never encountered any assertions is because I never use pyprof with dlprof enabled i.e. `config.capture_input_ops == True`.

Now that there is clarity on what ensued @debermudez can you fix this. @alexmsettle once this is fixed, let's revisit this PR. Graph information is extremely useful and it would be a great feature. I wholeheartedly apologize again for the misunderstanding and that some of the work will have to be redone. Hopefully not a lot and I can help with that.
